resumable: clarify handling of conflicts and retry policies

danielresnick opened the pull request:
See #3275 for discussion.

danielresnick committed:
resumable: clarify handling of conflicts and retry policies

See #3275 for discussion.

danielresnick committed:
Update changelog.

guoye-zhang commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -577,10 +579,9 @@ If the client received a final response with the `Upload-Complete: ?0` header fi
-- For a `5xx (Server Error)` status code, the client MAY automatically attempt upload resumption by retrieving the current offset ({{offset-retrieving}}).
-
-If no final response was received at all due to connectivity issues, the client MAY automatically attempt upload resumption by retrieving the current offset ({{offset-retrieving}}).
+- For a `409 (Conflict)` status code, the client SHOULD attempt to resume the upload by using the offset from the `Upload-Offset` response header field.
```
I think it's better to keep this flow the same as 5xx: recommending doing an offset retrieval before resuming. Otherwise this creates a new code path, and the text in "Upload Append" about picking the offset is also inconsistent with the recommendation here.

danielresnick replied:
Updated, how does this look?  It feels wrong to completely omit the option to leverage the Upload-Offset from the response since it's called out as a MUST for the server to return it just below

guoye-zhang commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -653,6 +654,56 @@ Content-Type: application/json
 }
 ~~~
 
+C) The following example shows a client attempting to resume an upload with an incorrect offset. The server rejects the request with a `409 (Conflict)` status code and includes the correct offset in the `Upload-Offset` header field. The client then retries the request with the correct offset.
```
The previous examples were all happy paths, and my preference is to keep it that way. If we want to include examples for exceptional cases, we should discuss what are the most common failures and why this particular case needs highlighting.

guoye-zhang reviewed (commented):
Thanks for opening this PR, the text looks good. Left a few minor comments.

danielresnick committed:
- Improved wording throughout.

- Addressed feedback around inconsistencies (HAVE vs MUST).

- Removed example 409 conflict recovery.

danielresnick committed:
Updated changelog to reflect latest revision.

Acconut commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -686,9 +687,15 @@ Resumable uploads, as defined in this document, do not permit uploading represen
+
+2.  **Pessimistic Locking**: The server processes requests for a given upload resource sequentially, effectively creating an exclusive lock on that resource. A new request is only processed after the previous one completes. This can be simpler to implement but may lead to delays if a request hangs from the server's perspective.
+
+3.  **Optimistic Concurrency Control**: By maintaining the upload's state in a strongly consistent datastore, the server can atomically check if the `Upload-Offset` in a request matches the resource's current state. If they do not match, the server can reject the request with a `409 (Conflict)` status code.
```
I'm not sure if this is enough to guarantee issue-free uploads. What if the server receives two requests for the same (correct) offset? They both match, but which content is appended to the upload resource? Did you mean to say that the content should be appended in one transaction together with the offset check? But that would make this the same as the pessimistic locking, I think.

smatsson replied:
I agree with Acconut here. To me approach 2 and 3 seem very similar in how they would work in practice. The server would receive two requests with the same offset (why though, really?) and one is slightly faster than the other i.e. it will get the exclusive lock and the thus the second request would fail. I think the wording of point 2 hints that the requests are queued and processed after each other which I guess is not the case (the offset would most likely never match for the second request)

danielresnick committed:
Update concurrency section & offset retrieval based on feedback.

Acconut commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -686,9 +687,15 @@ Resumable uploads, as defined in this document, do not permit uploading represen
+
+2.  **Pessimistic Locking**: The server processes requests for a given upload resource sequentially, effectively creating an exclusive lock on that resource. A new request is only processed after the previous one completes. This can be simpler to implement but may lead to delays if a request hangs from the server's perspective.
+
+3.  **Optimistic Concurrency Control**: The server detects concurrent modifications by atomically checking the resource's state before applying changes. If a conflict is detected, the server rejects the request with a 409 (Conflict) status code, ensuring that only one of multiple parallel requests can succeed.
```
I'm wondering how this would be achieved in practice on the server-side. Does the server read the entire request content first and then tries to apply it in one atomic operation that checks the offset and appends the content? The offset would then be validated after receiving the request content, not when the headers are sent. Not a problem, but a different angle to the task.

Small suggestion:

```suggestion
3.  **Optimistic Concurrency Control**: The server handles concurrent modifications by checking the resource's state and applying changes together in one atomic operation. If a conflict is detected, the server rejects the request with a 409 (Conflict) status code, ensuring that only one of multiple parallel requests can succeed. Upload cancellation is implemented through a similar atomic operation.
```

GrantGryczan replied:
I personally don't see the point in approach 3 at all. It seems like a bad idea to prioritize an older request over a newer one by rejecting the newer one, considering the client may no longer know about the older one, and this restricts the client from controlling it.

danielresnick committed:
Merge branch 'main' into patch-1

GrantGryczan commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -578,10 +580,9 @@ If the client received a final response with the `Upload-Complete: ?0` header fi
-- For a `5xx (Server Error)` status code, the client MAY automatically attempt upload resumption by retrieving the current offset ({{offset-retrieving}}).
-
-If no final response was received at all due to connectivity issues, the client MAY automatically attempt upload resumption by retrieving the current offset ({{offset-retrieving}}).
+- For a `409 (Conflict)` status code, the client SHOULD attempt to resume the upload by either retrieving the current offset ({{offset-retrieving}}) or using the value from the response's `Upload-Offset` header field.
```
See my concern here: https://github.com/httpwg/http-extensions/issues/3275#issuecomment-3990724220

But I think that's an underlying design concern rather than a concern with your specific change towards that design, so feel free to resolve this as off-topic.

Acconut replied:
I think it's sufficient to tie the retries to the presence of `Upload-Complete: ?0` in this case. It would indicate that the 409 comes from the resumable upload layer.

```suggestion
- For a `409 (Conflict)` status code with the `Upload-Complete: ?0` header field, the client SHOULD attempt to resume the upload by either retrieving the current offset ({{offset-retrieving}}) or using the value from the response's `Upload-Offset` header field.
```

danielresnick committed:
Merge branch 'main' into patch-1

danielresnick commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -686,9 +687,15 @@ Resumable uploads, as defined in this document, do not permit uploading represen
+3.  **Optimistic Concurrency Control**: The server detects concurrent modifications by atomically checking the resource's state before applying changes. If a conflict is detected, the server rejects the request with a 409 (Conflict) status code, ensuring that only one of multiple parallel requests can succeed.
 
-Since implementing this approach is not always technically possible or feasible, other measures can be considered as well. A simpler approach is that the server only processes a new request to retrieve the offset ({{offset-retrieving}}), append representation data ({{upload-appending}}), or cancellation ({{upload-cancellation}}) once all previous requests have been processed. This effectively imp ...
+Servers SHOULD choose a strategy that best fits their architecture while fulfilling the requirements of this section. Regardless of the chosen strategy, clients SHOULD be prepared to handle a `409 (Conflict)` response as a recoverable error, as described in {{upload-appending}}.
```
Yeah good point, removed

Acconut replied:
Did you try to push changes to this PR? I'm not seeing any yet :)
